### DISCUSSION TITLE
Public-RPS and proposal failure

### DISCUSSION THREAD

**Discussion by martijnlammers** [microsoft/CCF#3112] [Q&A] [answered]
I've create a single node network using cchost with only one initial activated member containing the default constitution.

> 
> ; config.ini
> enclave-file = ./libjs_generic.enclave.so.signed
> enclave-type = release
> consensus = cft
> rpc-address = 10.0.0.18:443
> node-address = 10.0.0.18:6600
> public-rpc-address = [public VM addr]:443
> ledger-dir = ./legder/
> node-cert-file = ./certificates/node_cert.pem
> 
> [start]
> network-cert-file = ./certificates/network_cert.pem
> member-info = "./certificates/Martijn_cert.pem,./certificates/Martijn_enc_pubk.pem"
> constitution = "./constitution/actions.js"
> constitution = "./constitution/validate.js"
> constitution = "./constitution/resolve.js"
> constitution = "./constitution/apply.js"

I'm currently attempting to change my JS application through proposal, however I'm running into two problems.
Firstly, I'm unable to access my node via my public address, only through my local address. 
Attempting to activate a member using:
_$ curl https://[public VM addr]:443/gov/ack/update_state_digest -X POST --cacert ../certificates/network_cert.pem --key ../certificates/Martijn_privk.pem --cert ../certificates/Martijn_cert.pem_

Outputs:

> curl: (60) SSL: no alternative certificate subject name matches target host name '[my pub addr]'
> More details here: https://curl.haxx.se/docs/sslcerts.html
> 
> curl failed to verify the legitimacy of the server and therefore could not
> establish a secure connection to it. To learn more about this situation and
> how to fix it, please visit the web page mentioned above.
> 

Port 443 is open.

Secondly, I am unable to apply my proposal to the running network via my local address. 

![image](https://user-images.githubusercontent.com/60792074/138254738-4fb4868e-4409-4c2f-901f-ee5e58708fc5.png)

Is there anything visible that I am doing wrong?

**Comment by eddyashton** [comment, marked as answer]
For the first part: It looks like the certs that are generated don't include the correct Subject Alternative Name (SAN) field that curl is looking for when it established a TLS connection. This should be derived from the `--public-rpc-address` option, but can also be explicitly specified with `--san` (see the [note](https://microsoft.github.io/CCF/main/operations/start_network.html?highlight=san#starting-the-first-node) below the diagram here). You can print the fields in the cert with
`$ openssl x509 -text -in ./certificates/node_cert.pem`

Some examples:

```
; config.ini
public-rpc-address = 8.8.8.8:443

$ openssl x509 -text -in ./certificates/node_cert.pem
...
            X509v3 Subject Alternative Name: 
                IP Address:8.8.8.8
...
```
```
; config.ini
public-rpc-address = ccf.example.com:443

$ openssl x509 -text -in ./certificates/node_cert.pem
...
            X509v3 Subject Alternative Name: 
                DNS:ccf.example.com
...
```
```
; config.ini
public-rpc-address = 8.8.8.8:443
san = dNSName:ccf2.example.com

$ openssl x509 -text -in ./certificates/node_cert.pem
...
            X509v3 Subject Alternative Name: 
                DNS:ccf2.example.com
...
```

Can you check what SAN is being inserted, and confirm it matches your public VM address (in particular, that it is correctly identified as a DNS/IP?)?

To rule out any other factors, you should first test that you can reach a simpler unauthenticated endpoint (ie `curl https://[public VM addr]:443/node/commit --cacert ../certificates/network_cert.pem`, which does not need user or member auth).


_EDIT: This theory on the second issue is wrong, see below*_
For the second issue, I think the problem is with how you specified the constitution in the `.ini` file. When we specify multiple files on the CLI, this is multiple instances to `--constitution`, but it looks like the TOML representation of that is a list, rather than multiple values. In other words, replace
```
constitution = "./constitution/actions.js"
constitution = "./constitution/validate.js"
constitution = "./constitution/resolve.js"
constitution = "./constitution/apply.js"
```
with
```
constitution=["./constitution/actions.js", "./constitution/validate.js", "./constitution/resolve.js", "./constitution/apply.js"]
```
(Note: Not sure about this one - that's how our arg parser _writes_ this value when we ask it to generate a `.ini` file, but afaict it should also parse the same from multiple entries. I'll see if I can work out what's going on there)

_* I've tested this locally, and having multiple instances of `constitution` isn't a problem - I was still able to propose and vote for (and Accept) a proposal. Your error message suggests there's something missing from the constitution. What proposal function are you trying to call? The error includes some line numbers in the constitution (528 and 929), but they correspond to the concatenated constitution so can't be trivially looked up in the source. You could use the `read_ledger` tool to dump the contents of the constitution table, unescape that string, and look at the source there, or just manually concatenate the constitution files (single newline between each) and see if there's anything obviously odd about the JS code at those lines?_

**Reply by martijnlammers** [reply]
It appears my network isn't able to open the constitution files. I've ran into this problem using the constitution straight from /runtime_config/default/ in the repo, however copying them in my working directory and importing them separately ended fixing it. ( So I thought ) I'm absolutely clueless what could cause this as I am running it with root privileges. 

> ; config.ini
> enclave-file = ./libjs_generic.enclave.so.signed
> enclave-type = release
> consensus = cft
> rpc-address = 10.0.0.18:443
> node-address = 10.0.0.18:6600
> public-rpc-address = [vm public address]:443
> ledger-dir = ./legder/
> node-cert-file = ./certificates/node_cert.pem
> 
> [start]
> network-cert-file = ./certificates/network_cert.pem
> member-info = "./certificates/Martijn_cert.pem,./certificates/Martijn_enc_pubk.pem"
> constitution= ["./constitution/actions.js", "./constitution/validate.js", "./constitution/resolve.js", "./constitution/apply.js"]
> 
Output running cchost --config=config.ini 

> . . . 
> 2021-10-21T14:52:24.965409Z        100 [info ] ../src/host/ledger.h:788             | Recovered ledger entries up to 0, committed to 0
> 2021-10-21T14:52:24.966388Z        100 [info ] ../src/host/node_connections.h:179   | Listening for node-to-node on 10.0.0.18:6600
> 2021-10-21T14:52:24.966963Z        100 [info ] ../src/host/rpc_connections.h:75     | Listening for RPCs on 10.0.0.18:443
> Could not open file 

This got me thinking, that perhaps the permissions are messed up?

![image](https://user-images.githubusercontent.com/60792074/138305340-0d01950e-5a56-495e-bc30-593e931c7582.png)

So, I ended up changing these. This did however not fix the problem.

![image](https://user-images.githubusercontent.com/60792074/138305568-e4af12ed-d2f3-44aa-8949-d35d74f93fc6.png)



With regards to the other issue, I ended up running 
$ openssl x509 -text -in ./certificates/node_cert.pem 
as you mentioned, which outputs that my SAN address is my local address, which doesn't correspond with my VM's address.

![image](https://user-images.githubusercontent.com/60792074/138307192-75f33788-d060-4a5f-bddb-eb5d75bda206.png)

I am capable to send requests to this address locally, but I not aware whether it's possible for other VM's to connect to this custom address.

**Reply by eddyashton** [reply]
That's very weird, the line is just `Could not open file`? That should be printing a filename (eg `Could not open file ./constitution/apply.js`), if there was a permissions issue. That suggests it's got an empty string somehow, and is trying to open that. There shouldn't be any permissions issues with the constitution files - they only need to be read, and they have read permissions in your screenshots, in the repo, and in the install folder. So I'm stumped there.

The `10.0.*` addresses are generally only visible to the VNet. So if you have multiple VMs within the same VNet, they should be able to talk to each other as `10.0.*` the same as they talk to 'themselves'. You can test that by pinging or SSHing from another VM. I don't know why the public RPC address isn't populating the SAN field. This behaviour may not be the same on the `1.x` branch, are you running a `1.x` build? If so, you may need the `--domain` or `--san` argument.

**Reply by martijnlammers** [reply]
I am currently running CCF 1.0.13
I understand the confusion. I have experienced "Could not open file ./\<path\>/apply.js" before. I'm a little confused myself as well to why it's giving me an empty string back now.


I'm now attempting to reconstruct the issue by using the constitution straight from the repo again and this did ended up giving me the error I was previously having.
Config:
> . . . 
>  [start]
> network-cert-file = ./certificates/network_cert.pem
>  member-info = "./certificates/Martijn_cert.pem,./certificates/Martijn_enc_pubk.pem"
>  constitution=["\~/CCF/src/runtime_config/default/actions.js", "\~/CCF/src/runtime_config/default/validate.js", "\~/CCF/src/runtime_config/default/resolve.js", "\~/CCF/src/runtime_config/default/apply.js"]
> 


Ouput:
> . . . 
> 2021-10-21T16:09:41.411527Z        100 [info ] ../src/host/node_connections.h:179   | Listening for node-to-node on 10.0.0.18:6600
> 2021-10-21T16:09:41.411968Z        100 [info ] ../src/host/rpc_connections.h:75     | Listening for RPCs on 10.0.0.18:443
> Could not open file ~/CCF/src/runtime_config/default/actions.js

Permissions from repo: 
![image](https://user-images.githubusercontent.com/60792074/138316893-9cf662b8-b2e4-4d80-b9db-f4e868b9d25c.png)

**Reply by eddyashton** [reply]
Ah ha - the problem there is the use of `~`. This is a magic character that will get expanded to `/home/<your_user>/` in many contexts, but not by `cchost`. Can you replace that with the full path?

**Reply by achamayou** [reply]
Specifically, `~` is expanded by the shell.

**Reply by martijnlammers** [reply]
I was really hoping for that to be the answer, but unfortunately now I'm back at the empty string issue.

![image](https://user-images.githubusercontent.com/60792074/138320227-05067c8b-4453-46ae-87db-9afe18da2ab0.png)

Also, a side note whenever I'm running cchost without sudo privileges, it ends up unable to write as well, would these two resulting errors stem from thesame issue?

$ /opt/ccf/bin/cchost --config=config.ini 

> . . . 
> terminating with uncaught exception of type std::logic_error: Failed to write to file: /opt/ccf/bin/cchost.pid
> Stack trace (most recent call last):
> #12   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
> #11   Object "/opt/ccf/bin/cchost", at 0x40d43d, in 
> . . . 
> Aborted (Signal sent by tkill() 24377 1000)
> 2021-10-21T16:40:30.567881Z        100 [fatal] ../src/ds/stacktrace_utils.h:53      | Handled fatal signal 6
> Aborted (core dumped)

**Reply by eddyashton** [reply]
Ah, that means that `cchost` without sudo escalation is unable to write to the `/opt/ccf/bin` directory (as expected), but by default tries to create a `.pid` file there! I'll see why our tests don't hit that. You can work around it by adding a `node-pid-file = ./cchost.pid` line to the config (telling it where to write this file, writing it locally where it presumably has permissions).

EDIT - the reason our tests never see this is that they use `sandbox.sh` which always copies the `cchost` binary locally, and never tries to run it in `/opt/ccf/bin`.

**Reply by martijnlammers** [reply]
That did indeed fix the issue of not being able to write cchost.pid.
Seemingly it does still seem to have other permission issues.

$ ./ccf/bin/cchost --config=config.ini
> . . .
> 2021-10-21T17:02:39.268742Z        100 [fail ] ../src/host/tcp.h:338                | uv_tcp_bind failed on 10.0.0.18:443: permission denied
> terminating with uncaught exception of type std::runtime_error: TCP server listen failed
> Stack trace (most recent call last):
> #15   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
> #14   Object "./ccf/bin/cchost", at 0x40d43d, in 
> . . . 
> #0    Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7f6d2763318b, in gsignal
> Aborted (Signal sent by tkill() 26172 1000)
> 2021-10-21T17:02:39.278523Z        100 [fatal] ../src/ds/stacktrace_utils.h:53      | Handled fatal signal 6
> Aborted (core dumped)

I tried copying the while /opt/ccf/ folder into my working directory where I should have permissions, but that action ended up being futile.

**Reply by eddyashton** [reply]
I think the problem there is trying to bind to `443` as non-root. Any well-known port (0-1023) is generally managed by the OS, and not available to user applications. So either use an alternative port (a random large number), or run it as root if you need it to be HTTPS-standard.

**Reply by achamayou** [reply]
@mAlyanak 443 is a system port, you cannot bind to it if you aren't root. Either run the process as root, or choose a port above 1023.

**Reply by martijnlammers** [reply]
Oh yeah that's right, very sharp to notice. Totally forgot about that little detail haha.

I ended up getting the empty string issue again.

> $ sudo ./ccf/bin/cchost --config=config.ini
> 021-10-21T17:18:09.800113Z        100 [info ] ../src/host/node_connections.h:179   | Listening for node-to-node on 10.0.0.18:6600
> 2021-10-21T17:18:09.800731Z        100 [info ] ../src/host/rpc_connections.h:75     | Listening for RPCs on 10.0.0.18:443
> Could not open file 

I tried putting the const. seperate again now, reattempting to apply a proposal again with little success.
#config

> [start]
> network-cert-file = ./certificates/network_cert.pem
> member-info = "./certificates/Martijn_cert.pem,./certificates/Martijn_enc_pubk.pem"
> constitution = "/home/mar/CCF/src/runtime_config/default/actions.js"
> constitution = "/home/mar/CCF/src/runtime_config/default/validate.js"
> constitution = "/home/mar/CCF/src/runtime_config/default/resolve.js"
> constitution = "/home/mar/CCF/src/runtime_config/default/apply.js"
> 

#Start network
$ sudo /opt/ccf/bin/cchost --config=config.ini

#In a seperate terminal:
#Activating initial member
$ curl https://10.0.0.18:443/gov/ack/update_state_digest -X POST --cacert ../certificates/network_cert.pem --key ../certificates/Martijn_privk.pem --cert ../certificates/Martijn_cert.pem

$ scurl.sh https://10.0.0.18:443/gov/ack --cacert ./../certificates/network_cert.pem --signing-key ./../certificates/Martijn_privk.pem --signing-cert ./../certificates/Martijn_cert.pem --header "Content-Type: application/json" --data-binary '{"state_digest":"a9021ac46265b4158fad13ff41b0f1bd7d6f6c088930468a28d402af32e901db"}'

#Send JS app proposal
$ scurl.sh https://10.0.0.18:443/gov/proposals --cacert ../certificates/network_cert.pem --signing-key ../certificates/Martijn_privk.pem --signing-cert ../certificates/Martijn_cert.pem  --data-binary @set_js_app_proposal.json -H "content-type: application/json"

$ scurl.sh https://10.0.0.18:443/gov/proposals/f8c8f9c84cc38f87c1a08df9d365338cd3fb1b6c14cf8936d3028bc9bbcc68ae/ballots --cacert ../certificates/network_cert.pem --signing-key ../certificates/Martijn_privk.pem --signing-cert ../certificates/Martijn_cert.pem --data-binary @vote_accept.json -H "content-type: application/json"

Output: 

> {"ballot_count":1,"failure":{"reason":"Failed to apply(): TypeError: not a function","trace":"    at <anonymous> (public:ccf.gov.constitution[0]:528)\n    at apply (public:ccf.gov.constitution[0]:959)\n"},"proposal_id":"f8c8f9c84cc38f87c1a08df9d365338cd3fb1b6c14cf8936d3028bc9bbcc68ae","proposer_id":"0eb426bc7351c937b03e81d23de9a17ea333fc650c5020dd57cf7a5de6df42ba","state":"Failed","votes":{"0eb426bc7351c937b03e81d23de9a17ea333fc650c5020dd57cf7a5de6df42ba":true}}

**Reply by eddyashton** [reply]
I think what's happening here is that you're using the `.js` scripts from the `main` branch of the repo, against a CCF node running `1.0.13`. I dumped the constitution from a local instance*, and sure enough line 528 is inside the `set_js_app` apply function, but it is calling `modulesMap.clear()` which doesn't exist in `1.0.13`. The fix here is to use the governance scripts from `1.0.13`, not from `main`.

\* In case it's useful to someone in future, I did that with `python ../python/ccf/read_ledger.py workspace/content_types_auth_0/0.ledger/ --uncommitted -t ".*constitution.*" | grep -A2 'table "public:ccf.gov.constitution" (1 write)' | tail -n1 | sed 's/\\n/\n/g; s/\\"/"/g' > ./constitution.js`. That's "call read_ledger.py, reading just the constitution table, grep out the constitution line, sed to unescape its newlines and quotes".

**Reply by martijnlammers** [reply]
Ohh, yeah that did seem to be the case. Major thanks for the help.
